rfd/0095-teleport-key-agent-extension.md

```diff
@@ -0,0 +1,251 @@
+
+```bash
+$ tsh login --user=dev --proxy=proxy.example.com:3080
+Enter password for Teleport user dev:
+> Profile URL:        https://proxy.example.com:3080
+  Logged in as:       dev
+  Cluster:            root-cluster
+  Roles:              dev
+  Logins:             dev, -teleport-internal-join
+  ...
+  
+$ tsh ssh -o "ForwardAgent local" server01
+### successfully connect
+
+<server01> $ ls ~/.tsh
+# empty
+
+<server01> $ tsh status
+> Profile URL:        https://proxy.example.com:3080
+  Logged in as:       dev
+  Cluster:            root-cluster
+  Roles:              dev
+  Logins:             dev, -teleport-internal-join
+  ...
+
+$ tsh ls
+Node Name Address        Labels                                                                             
+--------- -------------- --------------------------------
+server01  127.0.0.1:3022 arch=x86_64,cluster=root-cluster
+server02  127.0.0.1:3022 arch=x86_64,cluster=root-cluster
+
+tsh ssh -o "ForwardAgent local" server02
+### successfully connect
+```
+
+The remote session will only have access to the client certificates used for the current session, rather than the entirety of `~/tsh`. Additionally, these certificates will only be returned by the forwarded agent if the corresponding agent keys are also still available. This means that if the local uses `tsh logout` or changes profiles (`tsh login other-cluster`), the remote session will no longer have access to the `tsh` profile.
+
+```bash
+$ tsh ssh -o "ForwardAgent local" server01
+### successfully connect
+
+### switch tabs and `tsh logout` locally
+
+<server01> $ tsh status
+ERROR: Not logged in.
+```
+
+### Security
+
+#### SSH Agent Forwarding Risks
+
+SSH Agent forwarding comes with an inherent security risk. When a user does `ssh -A` or `tsh ssh -A`, their forwarded keys could be used by a user on the remote machine with OS permissions equal to or above the remote user. This is possible due to the way SSH channels are forwarded within a remote session, and this problem is shared by other protocols like X11 forwarding.
+
+The primary redeeming security principle for forwarded SSH channels is that forwarded keys cannot outlive the original session. This means that once the user who called `ssh -A` exits out of their session, no agent keys remain usable on the remote machine, and no sensitive data could have been exfiltrated to outlive the session. We'll call this security principle "session contingency" in this RFD.
+
+We can follow this same framework to forward a user's certificates and agent key without exposing their actual private key, only an interface to send the forwarded agent cryptographic challenges. This would allow users to forward their certificates and agent keys relatively safely to be used in a remote session and guarantee that malicious users cannot exfiltrate their `~/.tsh` and impersonate them. These forwarded keys could be used for any `tsh` and `tctl` request, or more generally, any Teleport API request.
```

Comment:
I would just suggest changing "we _can_ follow" to "we **must**", as under any circumstances we should not expose the user's private keys.